### PR TITLE
feat: Implement addExternalEndpointForGroup API

### PR DESCRIPTION
**Pull request by bhudnell**
Issue #, if available:

Description of changes:
- added `addExternalEndpointForGroup` API and corresponding unit/integ tests
- replaced `addExternalEndpointForUser` with `addExternalEndpointForGroup` in SWB
- removed `Boom` errors in favor of custom errors for service level functions related to `addExternalEndpointForGroup`

Checklist:

* [ ] Have you successfully deployed to an AWS account with your changes?
* [ ] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: workbench-core/example/app/src/routes/datasetRoutes.ts [hunk 1]
```diff
@@ -77,19 +88,47 @@ export function setUpDSRoutes(
...
+
+        if (validatedRequest.groupId) {
+          const { data } = await dataSetService.addDataSetExternalEndpointForGroup({
+            ...validatedRequest,
+            dataSetId: req.params.datasetId,
+            storageProvider: dataSetStoragePlugin,
+            groupId: validatedRequest.groupId,
+            authenticatedUser
+          });
+          res.status(201).send(data);
```
Comment (aws-tyler) [inline workbench-core/example/app/src/routes/datasetRoutes.ts:109]: lets do an early return here so we can avoid the else.

**Review by aws-tyler — APPROVED** [review]

File: workbench-core/datasets/src/dataSetService.ts [hunk 2]
```diff
@@ -606,4 +581,76 @@ export class DataSetService {
...
+      vpcId
+    } = request;
+
+    const { data: permissionsData } = await this._authzPlugin.getAccessPermissions({
+      dataSetId,
+      subject
+    });
+    if (!permissionsData.permissions.length) {
+      throw new NotAuthorizedError(
+        `Subject "${subject}" does not have permission to access dataset "${dataSetId}.`
```
Comment (kpark277) [inline workbench-core/datasets/src/dataSetService.ts:605]: I think there is a mismatch in concepts here; subject here would be 'dataset' or 'datasetPermission', not the user/group.

Comment (bhudnell) [reply workbench-core/datasets/src/dataSetService.ts:605]: I was following the inputs for `getDatasetPermissions` here: https://github.com/aws-solutions/solution-spark-on-aws/blob/develop/workbench-core/datasets/src/models/getAccessPermissionRequest.ts

File: workbench-core/datasets/src/dataSetService.test.ts [hunk 3]
```diff
@@ -576,24 +645,34 @@ describe('DataSetService', () => {
...
-      expect(Boom.isBoom(response, 400)).toBe(true);
-      expect(response.message).toEqual(
-        `'${mockExistingEndpointName}' already exists in '${mockDataSetWithEndpointId}'.`
-      );
+        })
+      ).rejects.toThrow(EndPointExistsError);
+    });
+
+    it('throws if the subject doesnt have permission to access the dataset', async () => {
+      jest
```
[the remark this answers is not in the document]
Comment (bhudnell) [reply workbench-core/datasets/src/dataSetService.test.ts:662]: They all have different mocked return values, so they need to stay in their respective tests. I did see that I duplicated a test, so removing that!

File: workbench-core/datasets/src/s3DataSetStoragePlugin.ts [hunk 4]
```diff
@@ -210,13 +214,20 @@ export class S3DataSetStoragePlugin implements DataSetsStoragePlugin {
...
+      );
+      return {
+        endPointArn: response.AccessPointArn!,
+        endPointAlias: response.Alias
+      };
+    } catch (error) {
+      if (error.name === 'AccessPointAlreadyOwnedByYou') {
+        throw new EndPointExistsError(error.message);
+      }
+      throw error;
```
[the remark this answers is not in the document]
Comment (bhudnell) [reply workbench-core/datasets/src/s3DataSetStoragePlugin.ts:229]: logging with the audit plugin is done in the public methods of the `DataSetService` class

**Review by aws-tyler — APPROVED** [review]

File: workbench-core/datasets/src/dataSetService.ts (hunk 2 above)
Comment (bhudnell) [reply workbench-core/datasets/src/dataSetService.ts:611]: The inputs have changed to use `identity` over `subject` so I think we're back on the same page in terms of concepts

File: solutions/swb-reference/src/services/dataSetService.ts [hunk 5]
```diff
@@ -48,9 +48,8 @@ export class DataSetService implements DataSetPlugin {
   public addDataSetExternalEndpoint(
     request: DataSetExternalEndpointRequest
   ): Promise<DataSetAddExternalEndpointResponse> {
-    return this._workbenchDataSetService.addDataSetExternalEndpointForUser({
+    return this._workbenchDataSetService.addDataSetExternalEndpointForGroup({
```
Comment (DaveMadsen) [inline solutions/swb-reference/src/services/dataSetService.ts:51]: do they only accept a group here? never a user? If they expose `identityType`, you should check that and call the appropriate method based on type.

File: workbench-core/datasets/src/dataSetService.test.ts [hunk 6]
```diff
@@ -548,8 +550,75 @@ describe('DataSetService', () => {
...
+      await expect(
+        dataSetService.addDataSetExternalEndpointForGroup({
+          dataSetId: mockDataSetWithEndpointId,
+          externalEndpointName: mockExistingEndpointName,
+          storageProvider: s3Plugin,
+          groupId: mockGroupId,
+          authenticatedUser: mockAuthenticatedUser,
+          externalRoleName: mockRoleArn
+        })
+      ).rejects.toThrow(EndPointExistsError);
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/dataSetService.test.ts:599]: `toThrowError`
Also consider verifying the error message.

File: workbench-core/datasets/src/dataSetService.test.ts [hunk 7]
```diff
@@ -548,8 +550,75 @@ describe('DataSetService', () => {
...
+      await expect(
+        dataSetService.addDataSetExternalEndpointForGroup({
+          dataSetId: mockDataSetWithEndpointId,
+          externalEndpointName: mockExistingEndpointName,
+          storageProvider: s3Plugin,
+          groupId: mockGroupId,
+          authenticatedUser: mockAuthenticatedUser,
+          externalRoleName: mockRoleArn
+        })
+      ).rejects.toThrow(NotAuthorizedError);
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/dataSetService.test.ts:616]: `toThrowError`
Also consider verifying the error message.

File: solutions/swb-reference/src/services/dataSetService.ts (hunk 5 above)
Comment (bhudnell) [reply solutions/swb-reference/src/services/dataSetService.ts:51]: @kpark277 SWB is only using `addDataSetExternalEndpointForGroup()` right?

**Review by aws-tyler — APPROVED** [review]

File: workbench-core/datasets/src/dataSetService.test.ts [hunk 8]
```diff
@@ -584,24 +653,17 @@ describe('DataSetService', () => {
...
-        });
-      } catch (err) {
-        response = err;
-      }
-      expect(Boom.isBoom(response, 400)).toBe(true);
-      expect(response.message).toEqual(
-        `'${mockExistingEndpointName}' already exists in '${mockDataSetWithEndpointId}'.`
-      );
+        })
+      ).rejects.toThrow(EndPointExistsError);
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/dataSetService.test.ts:666]: `toThrowError`
Also consider verifying the error message.

File: workbench-core/datasets/src/dataSetService.ts [hunk 9]
```diff
@@ -289,14 +295,15 @@ export class DataSetService {
 
       if (
         !targetDS.externalEndpoints ||
-        !_.find(targetDS.externalEndpoints, (ep) => ep === externalEndpointId)
-      )
+        !_.find(targetDS.externalEndpoints, (ep) => ep === targetEndpoint.name)
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/dataSetService.ts:298]: why this change? Have we been doing it wrong all this time?

File: workbench-core/datasets/src/dataSetService.ts [hunk 10]
```diff
@@ -289,14 +295,15 @@ export class DataSetService {
...
+      ) {
         return;
+      }
 
       await storageProvider.removeExternalEndpoint(targetEndpoint.name, targetDS.awsAccountId!);
 
       targetDS.externalEndpoints = _.remove(targetDS.externalEndpoints, (endpoint) => {
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/dataSetService.ts:305]: this code looks like it was always wrong. `_.remove()` returns the elements _removed_ from the array, not the remaining elements.

File: workbench-core/datasets/src/models/addDataSetExternalEndpoint.ts [hunk 11]
```diff
@@ -3,19 +3,21 @@
  *  SPDX-License-Identifier: Apache-2.0
  */
 
-import { AuthenticatedUser } from '@aws/workbench-core-authorization';
+import { AuthenticatedUser, IdentityType } from '@aws/workbench-core-authorization';
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/models/addDataSetExternalEndpoint.ts:6]: Do not import `authenticatedUser` from authorization in the `models/` as it tightly couples implementations. Better to redefine our expected `authenticatedUser` so we can eventually refactor to the point that `DataSets` can be consumed without the `@aws/workbench-core-authorization` dependency.

Comment (DaveMadsen) [reply workbench-core/datasets/src/models/addDataSetExternalEndpoint.ts:6]: I am fine with `DataSets` defining its own `authenticatedUser` object. It would also be okay to pull it from `base`, but it may be a bit difficult to discover there.

File: workbench-core/datasets/src/dataSetService.ts (hunk 10 above)
Comment (bhudnell) [reply workbench-core/datasets/src/dataSetService.ts:305]: Interesting. I think we need to take another look at our unit/integ tests for datasets because I feel like this should have been caught by them.

File: workbench-core/datasets/src/s3DataSetStoragePlugin.test.ts [hunk 12]
```diff
@@ -1043,6 +1044,40 @@ describe('S3DataSetStoragePlugin', () => {
...
+        plugin.addExternalEndpoint({
+          name,
+          path,
+          externalEndpointName,
+          ownerAccountId: mockAwsAccountId,
+          accessLevel: 'read-only',
+          externalRoleName: externalRoleArn,
+          kmsKeyArn
+        })
+      ).rejects.toThrow(EndPointExistsError);
```
Comment (DaveMadsen) [inline workbench-core/datasets/src/s3DataSetStoragePlugin.test.ts:1062]: `toThrowError`

I was also just reading suggests that say you should do 
`.toThrow(<error type).toThrow('error message')

File: workbench-core/datasets/src/models/addDataSetExternalEndpoint.ts (hunk 11 above)
Comment (bhudnell) [reply workbench-core/datasets/src/models/addDataSetExternalEndpoint.ts:6]: Wont that put us at risk of things breaking if the implementation of either of these changes? Could a solution be to re-export external package public facing interfaces in the index file? So no other package is needed

File: workbench-core/example/app/src/routes/datasetRoutes.ts [hunk 13]
```diff
@@ -81,19 +92,57 @@ export function setUpDSRoutes(
   router.post(
     '/datasets/:datasetId/share',
     wrapAsync(async (req: Request, res: Response) => {
-      if (req.params.datasetId.match(uuidWithLowercasePrefixRegExp(dataSetPrefix)) === null) {
```
Comment (DaveMadsen) [inline workbench-core/example/app/src/routes/datasetRoutes.ts:84]: Don't remove this. you are using `req.params.datasetId` later in your implementation (which is the right thing to do!), but you removed its validation!

File: workbench-core/example/app/src/routes/datasetRoutes.ts [hunk 14]
```diff
@@ -122,19 +171,26 @@ export function setUpDSRoutes(
   router.post(
     '/datasets/:datasetId/presignedUpload',
     wrapAsync(async (req: Request, res: Response) => {
-      if (req.params.datasetId.match(uuidWithLowercasePrefixRegExp(dataSetPrefix)) === null) {
```
Comment (DaveMadsen) [inline workbench-core/example/app/src/routes/datasetRoutes.ts:125]: same issue. validation of the parameter is removed, but then the parameter is used later.

File: workbench-core/example/app/src/routes/datasetRoutes.ts [hunk 15]
```diff
@@ -81,19 +92,57 @@ export function setUpDSRoutes(
...
     '/datasets/:datasetId/share',
     wrapAsync(async (req: Request, res: Response) => {
-      if (req.params.datasetId.match(uuidWithLowercasePrefixRegExp(dataSetPrefix)) === null) {
-        throw Boom.badRequest('datasetid request parameter is invalid');
+      try {
+        const validatedRequest = validateAndParse<CreateExternalEndpoint>(
+          CreateExternalEndpointParser,
+          req.body
+        );
+        const authenticatedUser = validateAndParse<AuthenticatedUser>(
```
Comment (DaveMadsen) [inline workbench-core/example/app/src/routes/datasetRoutes.ts:100]: do we need to do this everywhere?

Comment (bhudnell) [reply workbench-core/example/app/src/routes/datasetRoutes.ts:100]: Not a bad idea imo

File: workbench-core/example/app/src/routes/datasetRoutes.ts [hunk 16]
```diff
@@ -151,11 +207,15 @@ export function setUpDSRoutes(
   router.get(
     '/datasets/:datasetId',
     wrapAsync(async (req: Request, res: Response) => {
-      if (req.params.datasetId.match(uuidWithLowercasePrefixRegExp(dataSetPrefix)) === null) {
```
Comment (DaveMadsen) [inline workbench-core/example/app/src/routes/datasetRoutes.ts:154]: don't remove validation for parameters you don't validate elsewhere.

File: workbench-core/example/app/src/routes/datasetRoutes.ts [hunk 17]
```diff
@@ -172,16 +232,15 @@ export function setUpDSRoutes(
   router.delete(
     '/datasets/:datasetId',
     wrapAsync(async (req: Request, res: Response) => {
-      if (req.params.datasetId.match(uuidWithLowercasePrefixRegExp(dataSetPrefix)) === null) {
```
Comment (DaveMadsen) [inline workbench-core/example/app/src/routes/datasetRoutes.ts:175]: don't remove.

File: workbench-core/example/app/src/routes/datasetRoutes.ts (hunk 13 above)
Comment (bhudnell) [reply workbench-core/example/app/src/routes/datasetRoutes.ts:84]: This is dead code, since authZ's static route config wont let a param that doesnt match get to the route handler

File: workbench-core/example/infrastructure/integration-tests/support/utils/regExpressions.ts [hunk 18]
```diff
@@ -2,8 +2,11 @@
  *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  *  SPDX-License-Identifier: Apache-2.0
  */
-import { resourceTypeToKey, uuidWithLowercasePrefixRegExp } from '@aws/workbench-core-base';
+import { uuidWithLowercasePrefixRegExp } from '@aws/workbench-core-base';
+import { endPointPrefix, dataSetPrefix } from '@aws/workbench-core-example-app/lib/configs/constants';
 
-export const envUuidRegExp: RegExp = uuidWithLowercasePrefixRegExp(resourceTypeToKey.environment);
```
Comment (DaveMadsen) [inline workbench-core/example/infrastructure/integration-tests/support/utils/regExpressions.ts:7]: why is this being removed?

Comment (bhudnell) [reply workbench-core/example/infrastructure/integration-tests/support/utils/regExpressions.ts:7]: We arent integrating environments into the example app so its not necessary

Comment (bhudnell) [reply workbench-core/example/infrastructure/integration-tests/support/utils/regExpressions.ts:7]: Was originally here as a relic of the copy of how SWB set up their integ test infrastructure

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/create.test.ts [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+  });
+
+  describe('AddExternalEndpointForUser', () => {
+    let dataset: Dataset;
+    let userId: string;
+    let identityType: string;
+
+    beforeAll(async () => {
+      identityType = 'USER';
+      const { data } = await adminSession.resources.users.create({
```
Comment (DaveMadsen) [inline workbench-core/example/infrastructure/integration-tests/tests/datasets/create.test.ts:45]: I have found this works better in the outer most `beforeAll()` block. That way, you can create one user and utilize it throughout the test suite.

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/delete.test.ts [hunk 20]
```diff
@@ -26,7 +26,7 @@ describe('datasets delete integration test', () => {
     user = {
       firstName: 'Test',
       lastName: 'User',
-      email: `success+create-user-${uuidv4()}@simulator.amazonses.com`
+      email: `success+delete-dataset-${uuidv4()}@simulator.amazonses.com`
```
Comment (DaveMadsen) [inline workbench-core/example/infrastructure/integration-tests/tests/datasets/delete.test.ts:29]: lol. good catch!

File: workbench-core/example/infrastructure/integration-tests/tests/datasets/create.test.ts (hunk 19 above)
Comment (bhudnell) [reply workbench-core/example/infrastructure/integration-tests/tests/datasets/create.test.ts:45]: This is currently the only block that uses it. If another block does then it can be extracted.

File: workbench-core/example/infrastructure/package.json [hunk 21]
```diff
@@ -31,7 +31,7 @@
     "pkg-json-lint": "npmPkgJsonLint -c ../../../.npmpackagejsonlintrc.json .",
     "sort-package-json": "sort-package-json package.json",
     "test": "rushx test:only",
-    "test-cdk-deploy": "rush build && rushx compile && cdk deploy --all --require-approval never --outputs-file src/config/testEnv.json",
+    "test-cdk-deploy": "rush build -t @aws/workbench-core-example-infrastructure && rushx compile && cdk deploy --all --require-approval never --outputs-file src/config/testEnv.json",
```
Comment (DaveMadsen) [inline workbench-core/example/infrastructure/package.json:34]: will this ensure that ALL changes are picked up and packaged into the lambda handler? Or will it just rebuild the infrastructure package and potentially miss some things?

**Review by DaveMadsen — COMMENTED** [review]
There are some changes which should be made.
I'm not going to hit Request Changes as I need to run to a doctor's appointment and I don't want to hold you up if you can find another reviewer while I'm out.

File: workbench-core/example/infrastructure/package.json (hunk 21 above)
Comment (bhudnell) [reply workbench-core/example/infrastructure/package.json:34]: This builds the example infrastructure package and all dependencies (and dependencies of dependencies)

File: workbench-core/datasets/src/dataSetService.ts (hunk 10 above)
Comment (DaveMadsen) [reply workbench-core/datasets/src/dataSetService.ts:305]: what was the verdict here? Are we updating the dataset with the endpoints removed instead of the endpoints remaining?

**Review by DaveMadsen — APPROVED** [review]

File: workbench-core/datasets/src/dataSetService.ts (hunk 10 above)
Comment (bhudnell) [reply workbench-core/datasets/src/dataSetService.ts:305]: I'll take a look in my next ticket to double check things are working as intended. Want to get this PR merged
